src/service.py

```diff
@@ -0,0 +1,156 @@
+                    apt.DebianRepository(
+                        enabled=True,
+                        repotype="deb",
+                        uri="https://ppa.launchpadcontent.net/tphan025/ppa/ubuntu/",
+                        # TODO: depends on the series of the charm unit
+                        # set the release accordingly
+                        release="jammy",
+                        groups=["main"],
+                    )
+                )
+                apt.import_key(PPA_GPG_KEY_ID)
+            # Install the apt package
+            apt.update()
+            # Dependencies
+            apt.add_package("openjdk-17-jre")
+            # Jenkins agent package from PPA
+            apt.add_package(AGENT_PACKAGE_NAME)
+        except (apt.PackageError, apt.PackageNotFoundError) as exc:
+            raise PackageInstallError("Error installing the agent package") from exc
+
+    def restart(self) -> None:
+        """Start the agent service.
+
+        Raises:
+            ServiceRestartError: when restarting the service fails
+        """
+        # Render template and write to appropriate file if only credentials are set
+        if credentials := self.state.agent_relation_credentials:
+            with open("templates/jenkins_agent_env.conf.j2", "r", encoding="utf-8") as file:
+                template = Template(file.read())
+            # fetch credentials and set them as environments
+            environments = {
+                "JENKINS_TOKEN": credentials.secret,
+                "JENKINS_URL": credentials.address,
+                "JENKINS_AGENT": self.state.agent_meta.name,
+            }
+            # render template
+            rendered = template.render(environments=environments)
+            # Ensure that service conf directory exist
+            config_dir = Path(SYSTEMD_SERVICE_CONF_DIR)
+            config_dir.mkdir(parents=True, exist_ok=True)
+            # Write the conf file
+            logger.info("Rendering agent configuration")
+            logger.debug("%s", environments)
+            # file name (override.conf) is important for the service to import envvars
+            self._render_file(f"{config_dir.resolve().as_posix()}/override.conf", rendered, 0o644)
+        try:
+            systemd.service_restart(AGENT_SERVICE_NAME)
+        except systemd.SystemdError as exc:
+            raise ServiceRestartError("Error starting the agent service") from exc
+
+    def stop(self) -> None:
+        """Stop the agent service."""
+        try:
+            systemd.service_stop(AGENT_SERVICE_NAME)
+        except systemd.SystemdError:
+            # TODO: do we raise exception here?
+            logger.debug("service %s failed to stop", AGENT_SERVICE_NAME)
+
+    def readiness_check(self) -> bool:
+        """Check whether the service was correctly started.
+
+        Returns:
+            bool: indicate whether the service was started.
+        """
+        time.sleep(READINESS_CHECK_DELAY)
```

Comment:
May I ask what the delay might be for? Perhaps this might be more suitable when Jenkins was started but not when calling for readiness check.

Reply:
I put this in place simply because most of the agent logic was moved to a systemd service, and after running`systemd.service_restart` we can't be sure whether the agent started properly. So I added a delay to allow the service to "stabilize" so that we can know it's up and running.

I also improved the logic for this check to do it at an interval and errors out when reaching the timeout.